### ISSUE TITLE
[Question] How can I run from Team City, only certain canopy tests, that have a given attribute?

### ISSUE THREAD

**Issue by adalenthya** [lefthandedgoat/canopy#365]
I want to improve the way I am currently running my Canopy tests.
Currently I have a SmokeTests.fs file, where I made a selection of smoke tests from all other *.fs file that contain my functionality tests.
But, I want to be able to run my Canopy tests from Team City, by running only those particular tests that have an attribute like **[smoke tests]** or **[acceptance tests]**.
Because I find it hard to maintain many Test Suites like "SmokeTests.fs". 
Is there such a way to do this?
Thank you!

**Comment by lefthandedgoat** [comment]
canopy uses its own runner, which doesn't support attributes like an nUnit/MSTests runner would.

The best way that to do it is like in this starter kit: 
https://github.com/lefthandedgoat/canopyStarterKit


https://github.com/lefthandedgoat/canopyStarterKit/blob/master/canopyStarterKit/Tests.fs

This provide a function and 2 attributes, one describing the breadth of the test, and one describing the section of your site under test.  Then you can pass command line attributes to your tests and it will run on those attributed with your arguments.

Hope this helps!

**Comment by adalenthya** [comment]
How would I integrate canopyStarterKit project in my project? Seems to be doing exactly what I need.

**Comment by lefthandedgoat** [comment]
The simplest thing to do is just to copy paste the code.

You need to copy the code at the bottom here, as it does the work of only setting up the tests with the correct tags:

https://github.com/lefthandedgoat/canopyStarterKit/blob/master/canopyStarterKit/Tests.fs#L32-L45

This code provides your tags

https://github.com/lefthandedgoat/canopyStarterKit/blob/master/canopyStarterKit/Common.fs#L13-L34

This code uses a nuget package called Argu to bring arguments from the command line into a proper type,  its optional but really useful

https://github.com/lefthandedgoat/canopyStarterKit/blob/master/canopyStarterKit/Args.fs

This part will parse the arguments, register the appropriate tests and run them

https://github.com/lefthandedgoat/canopyStarterKit/blob/master/canopyStarterKit/Program.fs#L8-L22

**Comment by adalenthya** [comment]
Ok, and how would I let TeamCity know which tests to run? 
It seems that the way canopyStarterKit does it, is that my project will start to execute the all tests containing "smoke" and "full". But what if I want in the second run to execute only "smoke"? Would that imply changing my code to exclude "full"? 
If yes, then it means I need to maintain the test suite and commit changes to the project, on to server, quite often. 
Whereas, what I'd like to accomplish is to execute from TeamCity, each time, different sets of tests, without having to manipulate the code.

**Comment by lefthandedgoat** [comment]
You will have team city build the code, and run the resulting .exe with parameters like this
https://github.com/lefthandedgoat/canopyStarterKit/blob/master/build.fsx#L137-L139

You can setup fields in your team city build that default to things like "Full" and when you run the build, you can click the `...` next to the run button, and provide different values, like Smoke, or Login to those fields, that would then create a different set of arguments on the command line, which would then be parsed by Argu (optional) and passed into the registerTests code.

**Comment by lefthandedgoat** [comment]
Also, you can download the starter kit, and run the build, and tweak the values to get a feel for how it works.  Then copy in the important pieces into your code.

**Comment by adalenthya** [comment]
Yes, that is perfect! I will give it a try like you said, and then start to integrate in my project, this is exactly what I was trying to do. Thank you again for offering good advice and for taking the time to help!

**Comment by lefthandedgoat** [comment]
No problem! Hope it works out well for you!
